pkg/gather/clusterconfig/pod_network_connectivity_checks.go

```diff
@@ -0,0 +1,93 @@
+package clusterconfig
+
+import (
+	"context"
+	"encoding/json"
+
+	controlplanev1 "github.com/openshift/api/operatorcontrolplane/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"github.com/openshift/insights-operator/pkg/record"
+)
+
+// GatherPNCC collects a summary of failed PodNetworkConnectivityChecks.
+// The checks are requested via a dynamic client and
+// then unmarshaled into the appropriate structure.
+//
+// Resource API: podnetworkconnectivitychecks.controlplane.operator.openshift.io/v1alpha1
+// Docs for relevant types: https://pkg.go.dev/github.com/openshift/api/operatorcontrolplane/v1alpha1
+//
+// * Location in archive: config/podnetworkconnectivitychecks.json
+// * Id in config: pod_network_connectivity_checks
+func GatherPNCC(g *Gatherer, c chan<- gatherResult) {
+	gatherDynamicClient, err := dynamic.NewForConfig(g.gatherKubeConfig)
+	if err != nil {
+		c <- gatherResult{errors: []error{err}}
+		return
+	}
+	gatherKubeClient, err := kubernetes.NewForConfig(g.gatherProtoKubeConfig)
+	if err != nil {
+		c <- gatherResult{errors: []error{err}}
+		return
+	}
+
+	records, errors := gatherPNCC(g.ctx, gatherDynamicClient, gatherKubeClient.CoreV1())
+	c <- gatherResult{records: records, errors: errors}
+}
+
+func getUnsuccessfulChecks(entries []controlplanev1.LogEntry) []controlplanev1.LogEntry {
+	unsuccesseful := []controlplanev1.LogEntry{}
+	for _, entry := range entries {
+		if !entry.Success {
+			unsuccesseful = append(unsuccesseful, entry)
+		}
+	}
+	return unsuccesseful
+}
+
+func gatherPNCC(ctx context.Context, dynamicClient dynamic.Interface, coreClient corev1client.CoreV1Interface) ([]record.Record, []error) {
+	pnccListUnstruct, err := dynamicClient.Resource(pnccGroupVersionResource).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	jsonBytes, err := pnccListUnstruct.MarshalJSON()
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	pnccListStruct := controlplanev1.PodNetworkConnectivityCheckList{}
+	if err := json.Unmarshal(jsonBytes, &pnccListStruct); err != nil {
+		return nil, []error{err}
+	}
+
+	unsuccessful := []controlplanev1.LogEntry{}
+	for _, pncc := range pnccListStruct.Items {
+		unsuccessful = append(unsuccessful, getUnsuccessfulChecks(pncc.Status.Failures)...)
+		for _, outage := range pncc.Status.Outages {
+			unsuccessful = append(unsuccessful, getUnsuccessfulChecks(outage.StartLogs)...)
+			unsuccessful = append(unsuccessful, getUnsuccessfulChecks(outage.EndLogs)...)
+		}
+	}
+
+	reasons := map[string]map[string]int{}
+	for _, entry := range unsuccessful {
+		if _, reasonExist := reasons[entry.Reason]; reasonExist {
+			msgMap := reasons[entry.Reason]
+			if _, msgExist := msgMap[entry.Message]; msgExist {
```

Comment:
How about...

```go
if _, msgExist := msgMap[entry.Message]; msgExist {
    msgMap[entry.Message]++
    continue
}
msgMap[entry.Message] = 1
```

Reply:
Isn't `continue` considered a bad practice when it isn't strictly necessary? I would say that the readability is also worse than with a regular if-else because this isn't really a "happy path" situation with an error check or something like that. Both of the branches are a part of normal functionality.

Reply:
I don´t think so... when you put the `continue`, you mostly are saying: "there is nothing after this that you should worry about if it matches this condition"... IMO, it kind of reduces the cyclomatic complexity. My tendency is always to remove unnecessary `else` statements for that specific reason. If the `else` breaks the flow, better to use something that indicates it immediately.

Reply:
Statements like `continue` or `break` are practically equivalent to using `goto`, so I don't think they're ever the preferable way of solving a problem where two equally probably options exist. Go only suggests the omission of an `else` statement after a condition that either returns or terminates the runtime. To me this kind of abuse of flow control indicates some kind of a special case (we have a lot of code like this, where there is a `break`/`continue`/`return` in a condition), but it doesn't seem natural here. Using if-else clearly indicates that there are two options rather than one special condition (typically an error of some sort) and one normal/healthy path.

Reply:
I don't see any issue with the original Ivo's definition TBH.

Reply:
Nevermind, I've figured out that in Go, this entire syntax is unnecessary because you can safely increment non-existing keys in a Go map. Apparently that's the idiomatic Go solution to this problem. [source](https://stackoverflow.com/a/41851720/3043260)